### PR TITLE
Amqp connections are not getting disposed

### LINKED ISSUE #1770: DeviceClient.Dispose() does not close socket connections
I'm using Microsoft.Azure.Devices.Client 1.34.0 (the latest). It looks like this issue is back: https://github.com/Azure/azure-iot-sdk-csharp/issues/1098

To put it short: when DeviceClient.Dispose() is called, underlying socket connections remain open. Those are to our IoT Hub, port 6571.

This started to happen after updating from 1.33.1 and is wreaking havoc on our service. I see that 1.34.0 changelog has a bunch of "dispose"-related fixes, so one of those might be to blame here.

### PR DESCRIPTION
**Pull request by jamdavi**
Fixes #1770 by reverting regression introduced in [this change](https://github.com/Azure/azure-iot-sdk-csharp/commit/911923977c7ce279e2c73740d7160b3104231a3c#diff-305c99963184ff9db1704e4c76120b5c6cad3d959d8aa959f6787b25c8ec5381R550).

### CODE REVIEW

File: iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs [hunk 1]
```diff
@@ -548,17 +548,12 @@ protected override void Dispose(bool disposing)
             {
                 base.Dispose(disposing);
 
-                if (_disposed)
```
Comment (drwill-ms) [inline iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs:551]: Why is this a problem?

**Review by drwill-ms — APPROVED** [review]

File: iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs [hunk 2]
```diff
@@ -548,17 +548,12 @@ protected override void Dispose(bool disposing)
...
-                }
-
                 Logging.Info(this, $"{nameof(disposing)}");
 
                 if (disposing)
                 {
                     _closed = true;
-                    AmqpUnitManager.GetInstance().RemoveAmqpUnit(_amqpUnit);
+                    AmqpUnitManager.GetInstance()?.RemoveAmqpUnit(_amqpUnit);
                     _disposed = true;
```
Comment (azabbasi) [inline iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs:557]: if we never use the _disposed field, we should just remove it.

**Comment by azabbasi** [comment]
Fixes #1770

File: iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs (hunk 1 above)
Comment (abhipsaMisra) [reply iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs:551]: +1. We seem to set _disposed to true only after removing the amqp unit, so it shouldn't cause leaks, right?

File: iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs (hunk 2 above)
Comment (azabbasi) [reply iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs:557]: nvm. Talked to James offline and he explained it to me. 
Based on our conversation, we can still keep the if(_disposed) check if we move it above the base.Dispose() call.

**Review by azabbasi — APPROVED** [review]

File: iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs (hunk 1 above)
Comment (vinagesh) [reply iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs:551]: @abhipsaMisra - Are you not OOF :D Take rest!

Comment (azabbasi) [reply iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs:551]: _disposed is a protected field from the base class. That's what's throwing us all off

Comment (azabbasi) [reply iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs:551]: when we call base.Dispose(), the _disposed is getting set to false, then we bail out prematurely.

Comment (jamdavi) [reply iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs:551]: I moved the dispose to below the if statement to match the pattern of the other handlers.

**Review by vinagesh — APPROVED** [review]

**Comment by vinagesh** [comment]
Can we add anything to catch this ? E2E test? Not this PR but after this.

**Comment by jamdavi** [comment]
> Can we add anything to catch this before customers? E2E test? Not this PR but after this.

We could add some monitoring to long haul. To do this in an integration test we'd need to mark the `_amqpUnit` member as internal to check to see if it's still part of the `AmqpUnitManager` and if it's disposed. 

I'll put a PR out for that to see what everyone thinks.

**Comment by vinagesh** [comment]
> > Can we add anything to catch this before customers? E2E test? Not this PR but after this.
> 
> We could add some monitoring to long haul. To do this in an integration test we'd need to mark the `_amqpUnit` member as internal to check to see if it's still part of the `AmqpUnitManager` and if it's disposed.
> 
> I'll put a PR out for that to see what everyone thinks.

Awesome! We can see how the test looks but we should definitely add some monitoring to long haul as we did not catch it when we released this time and this seems to be something that we should cover as a basic functionality.

**Review by bikamani — APPROVED** [review]

**Review by abhipsaMisra — APPROVED** [review]
Everyone is on board! 😄

**Pull request merged**